File: ruby/matching-brackets/matching_brackets.rb
# frozen_string_literal: true

class Brackets
  OPEN_BRACKETS = '({['
  CLOSE_BRACKETS = ')}]'

  def self.paired?(string)
    bracket_stack = []

    string.chars.each do |char|
      bracket_stack.push(char) if OPEN_BRACKETS.include?(char)
      next unless CLOSE_BRACKETS.include?(char)

      if bracket_stack.empty? || CLOSE_BRACKETS.index(char) != OPEN_BRACKETS.index(bracket_stack.pop)
        return false
      end
    end

    bracket_stack.empty?
  end
end

File: python/matching-brackets/matching_brackets.py
def is_paired(input_string):
    OPEN_BRACKETS = "({["
    CLOSE_BRACKETS = ")}]"

    bracket_stack = []
    for char in input_string:
        if char in OPEN_BRACKETS:
            bracket_stack.append(char)
        if char not in CLOSE_BRACKETS:
            continue
        if len(bracket_stack) == 0 or CLOSE_BRACKETS.index(char) != OPEN_BRACKETS.index(bracket_stack.pop()):
            return False

    return len(bracket_stack) == 0
